### PR TITLE
[BEAM-3369] Allow any coder to specify its deterministic variant.

### PR DESCRIPTION
**Pull request by robertwb**
This is preferable to hard-coding a list in typecoders.py.
The sequence coders have all been updated to include this ability.

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @chamikaramj

File: sdks/python/apache_beam/coders/coders.py [hunk 1]
```diff
@@ -97,6 +97,16 @@ def is_deterministic(self):
     """
     return False
 
+  def as_deterministic_coder(self, step_label, error_message=None):
```
Comment (chamikaramj) [inline sdks/python/apache_beam/coders/coders.py]: Should the runner guarantee that only one of these will be used ? For example, can a runner encode from non-deterministic one and decode from the corresponding deterministic one. Probably we should tighten the spec more here.

Comment (robertwb) [reply sdks/python/apache_beam/coders/coders.py]: This is in the code where coders are being chosen, so there's no separate encode/decode side here, and arbitrary substitution can be done.

**Comment by chamikaramj** [comment]
LGTM. Thanks.

**Pull request merged**